### PR TITLE
support two level deep nested resources

### PR DESCRIPTION
**Pull request by mattma**
This is a new feature that enable two level deep nested resources including a new unit test for this feature. 

New api: `/api/{parents}/:parent_id/{objects}/:child_id/{objects}/:id`

Originally, I want to make it recursively walk the directory tree. I thought about a new API. 

 `/api/{parents}/:parent_0_id/{objects}/:parent_1_id/{objects}/:parent_2_id/{objects}:id`

The number variable here is the `level` which is being used in the [source code](https://github.com/eggjs/egg-rest/blob/master/app.js#L37). Then I realized that it will break the existing API, and may require to bump up the version to 2.0.0. Well, since the project is in its own infancy, I want to keep it in minimal. 

Two level deep of nesting is what I need it now. I think it should cover lots of different Use Cases. If we could merge this one in, it would satisfy majority users. 

@fengmk2 @popomore @atian25

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -43,9 +43,12 @@ function loadRestApis(app) {
         // nesting is supported, for only one layer at most, `/api/parents/:parent_id/objects/:id`
```
Comment (atian25) [inline app.js]: comment should also modify

Comment (mattma) [inline app.js]: @atian25 updated!

File: README.md [hunk 2]
```diff
@@ -92,15 +92,15 @@ method     | url                                                    | file path
...
+Nesting of two layer at most is supported.
 
 method     | url                                                                       | file path                         | controller name
 ---        | ---                                                                       | ---                               | ---
-**GET**    | `/doc/api/{parents}/:parent_id/{objects}?per_page={per_page}&page={page}` | `app/apis/{parents}/{objects}.js` | **index()**
-**GET**    | `/doc/api/{parents}/:parent_id/{objects}/:id`                             | `app/apis/{parents}/{objects}.js` | **show()**
-**POST**   | `/doc/api/{parents}/:parent_id/{objects}`                                 | `app/apis/{parents}/{objects}.js` | **create()**
-**PUT**    | `/doc/api/{parents}/:parent_id/{objects}/:id`                             | `app/apis/{parents}/{objects}.js` | **update()**
-**DELETE** | `/doc/api/{parents}/:parent_id/{objects}/:id[s]`                          | `app/apis/{parents}/{objects}.js` | **destroy()**
+**GET**    | `/doc/api/{parents}/:parent_id/{children}/:child_id/{objects}?per_page={per_page}&page={page}` | `app/apis/{parents}/{objects}.js` | **index()**
```
Comment (fengmk2) [inline README.md]: breaking change?

Comment (mattma) [inline README.md]: I do not think it is a breaking change. The current feature should not break by any mean.

One level deep nesting should work as expecting. 

What this PR do?

Add two level deep nesting feature by introducing `:child_id` query parameter by using all existing the mechanic. 

That is why I explain earlier in my intention without introducing any breaking change?

Comment (mattma) [inline README.md]: If you see any breaking existing behavior, I need to rework on the PR. That is not intention at all.

**Comment by mattma** [comment]
@atian25 @popomore Any comment? Good to be merged? :)

**Comment by atian25** [comment]
I had not idea of this plugin

**Comment by mattma** [comment]
@atian25 You are the 2nd highest committer in this plugin. :) thanks. 

@shaoshuai0102 Could you chip in with what you think? I am kind of needing this feature, please, help.

File: README.md (hunk 2 above)
Comment (popomore) [inline README.md]: The file path and the url should be matched, why the file path is not `apis/{parents}/{children}/{objects}.js`.

Comment (mattma) [inline README.md]: @popomore 

It is indeed in this format `apis/{parents}/{children}/{objects}.js`

I have updated the readme two days ago based on @fengmk2 's comment. 

[See here](https://github.com/eggjs/egg-rest/blob/feature/nested/README.md#nested-resources)

File: test/fixtures/rest/app/apis/users/posts/answer/index.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+'use strict';
+
+exports.index = function* () {
```
Comment (popomore) [inline test/fixtures/rest/app/apis/users/posts/answer/index.js]: @mattma Forget to comment. I mean `/api/users/3/posts/1/replies/2/answer` is matching to `apis/users/posts/replies/answer.js`. I can retrive the ids from `this.params.parent_ids` which is

```
{
  users: '3',
  posts: '1',
  replies: '2',
}
```

Comment (mattma) [inline test/fixtures/rest/app/apis/users/posts/answer/index.js]: @popomore You are absolutely right. It is the idea. The PR update should reflect exactly the concept that you listed above. Code updated.

Comment (popomore) [inline test/fixtures/rest/app/apis/users/posts/answer/index.js]: The testcase hasn't been modified, I think the filepath should be `apis/users/posts/replies/answer.js`. I don't know why `answer` is sibling of `replies`, not the child of `replies`.

Comment (mattma) [inline test/fixtures/rest/app/apis/users/posts/answer/index.js]: @popomore Updated the readme, now it is clearly for what I have done. The `apis/users/posts/replies/answer.js` is not supported at the moment. The maximum nesting layers are three with is `users`, `posts`, `replies`. `answer` was never supported. 

I do have a testing use case to test on 404 for `answer.js` though, please, checkout `rest.test.js`. Hopefully it makes senses to you.

Comment (popomore) [inline test/fixtures/rest/app/apis/users/posts/answer/index.js]: I got it.

**Comment by popomore** [comment]
LGTM

**Comment by popomore** [comment]
I am merging it and publishing with minor

**Comment by mattma** [comment]
@popomore Thanks for all your help. 

I am kind of needing this feature. Is there anyway that we could cut a minor release for `egg` core by any chance for this feature and other features which we could release at the moment? @fengmk2 @atian25

**Comment by popomore** [comment]
+ egg-rest@1.1.0

**Comment by popomore** [comment]
egg use semver `^` to depend on egg-rest, you can reinstall egg.

**Comment by mattma** [comment]
@popomore Awesome. Thanks!

**Pull request merged**